### ISSUE TITLE
Fix numpy requirement version bug

### ISSUE THREAD

**Issue by fandinomat** [radiocosmology/caput#87]
When installing caput the following error may occur:

```Could not find a version that satisfies the requirement numpy==1.15.1 (from versions: 1.11.3, 1.12.1, 1.13.0, 1.13.1, 1.13.3, 1.14.0, 1.14.1, 1.14.2, 1.14.3, 1.14.5, 1.15.0, 1.15.2, 1.15.4, 1.16.0, 1.16.2, 1.16.3)
  No matching distribution found for numpy==1.15.1```

I believe this could be solved by changing the requirement in `pyproject.toml` from `'numpy==1.15.1` to `'numpy>=1.15.1` ?

@anjakefala @ketiltrout @nritsche

**Comment by anjakefala** [comment]
@fandinomat Fascinating! I wonder when they removed it.

The issue is that there was a separate run-time bug that I encountered with numpy 1.16.x.

You can see it at the bottom here: https://travis-ci.org/radiocosmology/caput/jobs/578074799. It resulted in failed tests and a red build. I do not fully understand why we have that binary incompatibility with > 1.16.0

We can test a selection of numpy ranges and adjust it to less than 1.16 but higher than the lowest version that works? That should give pip more flexibility.

**Comment by ketiltrout** [comment]
Gah.  The joys of `pyproject.toml` files.

You have to set the `numpy` version to the _oldest_ version (or older) that any of your installed wheels use.  The `h5py-*.whl` that gets installed is likely the culprit, but I don't know how to figure out what version of numpy it was built against.

So, `numpy<=1.15.1` is probably what you want, not `>=`.

A possible option is to not use a pre-built `h5py` in the CI.  When we deploy to the telescope with ansible we pass `--no-binary=h5py` to `pip install` to avoid this sort of thing.

**Comment by ketiltrout** [comment]
Let me try again...

This is the "fun" part of the interaction between `pyproject.toml`, wheels, and cython.

There is a binary incompatibility between the `numpy` used to build one of the wheels Travis installed in the CI and the numpy that's been installed (which appears to be `1.15.4` not `1.16`).  The most likely culprit is the `h5py-*.whl` that gets installed (but I don't know that for sure).

So, really this needs be set to the version used to build `h5py-*.whl` (or whatever other binary package is causing the problem). I think that's hard to do: it's always going to be a moving target because we can't control when new builds of things like `h5py` get pushed to PyPI (even if we fix the `h5py` version).

`numpy` claims backwards compatibility in their ABI, meaning any version newer than the one used to build the wheels should work, even if older ones don't.

If we're just trying to fix the CI build, when deploying to the telescope we pass `--no-binary=h5py` to `pip install` to avoid using the pre-build package.  That could be done in the Travis script as well.

**Comment by ketiltrout** [comment]
I've been playing around with things a bit.

If I build `caput` using `numpy>=0.16` and then try using caput with `numpy<1.16` I get that error that Travis had:

> ValueError: numpy.ufunc size changed, may indicate binary incompatibility. Expected 216 from C header, got 192 from PyObject

With any `numpy>=0.1.6` it works fine.  So, essentially I think the problem with that Travis build was in `setup.py` which simply had `"numpy"` as a requirement, meaning the `numpy-1.15.4` that was already installed satisfied it. We should have `setup.py` should have had `"numpy>=1.16"`.

My conclusion is, if we bump the numpy requirement to `>=1.16` everywhere (`requirements.txt`, `pyproject.toml`, `setup.py`) so it's all consistently past the Numpy ABI change, I think it will work.
